rp2040-hal/src/clocks/mod.rs

```diff
@@ -22,21 +51,108 @@ impl ShareableClocks {
     }
 }
 
+const XOSC_MHZ: u32 = 12_000_000_u32;
+
+fn make_div<S: TryInto<Hertz<u32>>, F: TryInto<Hertz<u32>>>(
+    src_freq: S,
+    freq: F,
+) -> Result<u32, ()> {
+    let src_freq = u64::from(*src_freq.try_into().map_err(|_| ())?.integer());
+    let freq = (*freq.try_into().map_err(|_| ())?.integer()).into();
+    let div: u64 = (src_freq << 8).wrapping_div(freq);
+    Ok(div as u32)
+}
 /// Abstraction layer providing Clock Management.
 pub struct ClocksManager {
     clocks: CLOCKS,
     shared_clocks: ShareableClocks,
 }
 impl ClocksManager {
     /// Exchanges CLOCKS block against Self.
-    pub fn new(mut clocks_block: CLOCKS) -> Self {
+    pub fn new(mut clocks_block: CLOCKS, watchdog: &mut Watchdog) -> Self {
+        // Start tick in watchdog
+        watchdog.enable_tick_generation(XOSC_MHZ as u8);
+
+        // Disable resus that may be enabled from previous software
+        clocks_block.clk_sys_resus_ctrl.write_with_zero(|w| w);
+
         let shared_clocks = ShareableClocks::new(&mut clocks_block);
         ClocksManager {
             clocks: clocks_block,
             shared_clocks,
         }
     }
 
+    /// Initialize the clocks
+    pub fn init(
+        &self,
+        _: &CrystalOscillator<Stable>,
+        _: &PhaseLockedLoop<Locked, PLL_SYS>,
+        _: &PhaseLockedLoop<Locked, PLL_USB>,
+    ) {
+        // Configure clocks
+        // CLK_REF = XOSC (12MHz) / 1 = 12MHz
+        let mut ref_clock = self.ref_clock();
+        let div = make_div(12u32.MHz(), 12u32.MHz()).unwrap();
+        if div > ref_clock.get_div() {
+            ref_clock.set_div(div);
```

Comment:
The Pico SDK you copied this from had a comment explaining why div is conditionally set once and then set again after changing clocks, but you didn't copy it across.
I recommend that you include this note because the reason for doing this is not immediately obvious to someone reading this code.

Reply:
Yeah due to how the current code is structured and how the pac works I could not get an C SDK equivalent set_clock function to work. So I figured the next best thing is to unroll it for every clock so there is at least an easy way to enable them... 

I will add that comment back. But alll that stuff should be abstracted over in a proper function

Reply:
Here's the comment you added back

**// If increasing divisor, set divisor before source.**

Compare that with the C SDK comment

**// If increasing divisor, set divisor before source. Otherwise set source before divisor.
// This avoids a momentary overspeed when e.g. switching to a faster source and increasing divisor to compensate.**

The second line there is much more valuable that the first, because it's describing *why* the code is performing actions that way.
The code can help us understand _what_ and _how_, but it does not provide any insight as to _why_ that method is important.

Permalink to the C code, for reference
https://github.com/raspberrypi/pico-sdk/blob/bfcbefafc5d2a210551a4d9d80b4303d4ae0adf7/src/rp2_common/hardware_clocks/clocks.c#L55

We'll fix it up post merge, but please try to keep in mind that a person reading your code has less knowledge of the problem than you do, and it's in your best interest to help them understand what you're trying to do.